File: Sem3/Advanced Methods of Programming/MyInterpreter - JavaFX - Final Project/project/src/domain/statement/FileStatements/ReadStatement.java
package domain.statement.FileStatements;


import domain.ProgramState;
import domain.adt.AdtException;
import domain.adt.MyDictionaryInterface;
import domain.adt.MyPair;
import domain.expression.Expression;
import domain.expression.ExpressionException;
import domain.statement.StatementException;
import domain.statement.StatementInterface;

import java.io.BufferedReader;
import java.io.IOException;


public class ReadStatement implements StatementInterface
{
    private Expression expFileId;
    private String varName;


    public ReadStatement(Expression expFileId, String varName)
    {
        this.expFileId = expFileId;
        this.varName = varName;
    }


    @Override
    public ProgramState execute(ProgramState programState) throws StatementException
    {
        MyDictionaryInterface<Integer, MyPair> fileTable = programState.getFileTable();
        MyDictionaryInterface<String, Integer> symbolTable = programState.getSymbolTable();
        MyDictionaryInterface<Integer, Integer> heapTable = programState.getHeapTable();
        try
        {
            int uniqueKey = expFileId.evaluate(symbolTable, heapTable);
            MyPair pair = fileTable.lookup(uniqueKey);
            BufferedReader reader = (BufferedReader) pair.getSecond();
            int valueFromFile;
            String result = reader.readLine();

            if (result == null)
                valueFromFile = 0;
            else
                valueFromFile = Integer.parseInt(result);

            if (symbolTable.isDefined(varName))
                symbolTable.update(varName, valueFromFile);
            else
                symbolTable.add(varName, valueFromFile);
        }
        catch (ExpressionException | AdtException exception)
        {
            throw new StatementException(exception.getMessage());
        }
        catch (IOException exception)
        {
            throw new StatementException("Cannot read from BufferedReader.");
        }
        return null;
    }


    @Override
    public String toString()
    {
        return "ReadFile(" + expFileId + "," + varName + ')';
    }
}

File: Sem3/Advanced Methods of Programming/MyInterpreter - CSharp Console Application/MyInterpreter CSharp/domain/statement/ReadStatement.cs
using System;
using System.IO;
using MyInterpreter_CSharp.domain.adt;
using MyInterpreter_CSharp.domain.expression;

namespace MyInterpreter_CSharp.domain.statement
{
    public class ReadStatement : IStatement
    {
        private Expression _expFileId;
        private string _varName;

        public ReadStatement(Expression expFileId, string varName)
        {
            _expFileId = expFileId;
            _varName = varName;
        }

        public ProgramState Execute(ProgramState programState)
        {
            IMyDictionary<int, MyPair<string, TextReader>> fileTable = programState.FileTable;
            IMyDictionary<string, int> symbolTable = programState.SymbolTable;

            try
            {
                int uniqueKey = _expFileId.Eval(symbolTable);
                MyPair<string, TextReader> myPair = fileTable.Lookup(uniqueKey);
                TextReader textReader = myPair.Second;
                int valueFromFile;
                string result = textReader.ReadLine();

                if (result == null)
                    valueFromFile = 0;
                else
                    valueFromFile = int.Parse(result);

                if (symbolTable.IsDefined(_varName))
                    symbolTable.Update(_varName, valueFromFile);
                else
                    symbolTable.Add(_varName, valueFromFile);
            }
            catch (ExpressionException exception)
            {
                throw new StatementException(exception.Message + " " + _expFileId);
            }
            catch (AdtException exception)
            {
                throw new StatementException(exception.Message + " " + _varName);
            }

            return null;
        }

        public override string ToString()
        {
            return "ReadFile(" + _expFileId + ", " + _varName + ')';
        }
    }
}